### PR TITLE
discourse_migration: Complete the mailing list to Discouse mappings

### PR DESCRIPTION
**Pull request by tstellar**

### CODE REVIEW

**Review by tlattner — CHANGES_REQUESTED** [review]
Regarding cfe-users, I actually think that Clang should be a top-level category with sub-categories such as "users". I think we need help from Discourse to move Clang out from under LLVM Project. 

ClangD would probably go under Clang category? Or it would have it's own top-level category. We have many sub-projects of LLVM and I think we can either have them all under one category or break them out into their own. For example, MLIR has it's own top level category. Should Clangd have one? When does it make sense for that situation? 

I would have similar questions about libclc, libcxx-dev, lldb-dev, flang, openmp, parallel_libs_dev.

Docs -> no need to migrate/deprecated (We should have Documentation as a sub-category or top level though but it wouldn't need these archives)
GSOC-> no need to migrate/deprecated

llvm-announce -> We should have a top level "Announce" category at the top. I would also think Beginners should really be topic related and not a general category.

Where should releases go? Have it's own top level category?

**Comment by tstellar** [comment]
What about something like this:
- clang
  - dev
  - users
  - clangd
- libcxx
  - users
  - dev
  - libunwind
  - parallel-libs
- lldb
  - users
  - dev
- Runtimes
  - sanitizers
  - libclc
  - openmp
- flang
  - users
  - dev
- llvm
  - users
  - dev

**Comment by tstellar** [comment]
I forgot about lld, that probably deserves a top-level category too.  I thin it might be good to keep the llvm-project top-level category and put the releases sub-category there and maybe documentation too?  I don't really know what to do about the beginners category, to me that is more or less the same as the users category.

**Comment by joker-eph** [comment]
Users in clang exists for people who are really end-users of the tool and not developers. Beginner is more about people who are trying to get started with developing inside clang instead. The same logic may apply to other project I guess...

@tlattner : originally MLIR was also under LLVM project and we asked to move out so that we can have subcategories (Discourse limits to one level of nesting). But the current setup otherwise was very intentional, I'll ping Meike so that she can elaborate a bit on this maybe.

**Comment by Somae** [comment]
@joker-eph thank you for pining me.

We can of course make all of these things top level categories, but it will make the main page very crowded and hard to orient. It makes a lot of sense to me to create a few more top level categories, but I don't see the strict need for all of them.

Cases where I think we will have a benefit: If something, like MLIR did back then, has multiple subcategories (MLIR main category plus 3 subcategories). What is the purpose of the top level categories you are proposing? Usually you can open topics in it as well, and MLIR uses it for the main discussions.
Should the main category be the equivalent of the users category? Or dev? We could also establish the already existing LLVM top level category as the equivalent of either of those. If we have a clear vision here and still think everything should be a main category, we can make those changes of course, but I'd like us to have a plan what kind of discussions we want to facilitate where. Wild suggestion as example what I mean: We could even do a split between a "users" main category and a "dev" main category, which is the split for most and also a distinct set of users, probably, and then for the few bigger areas elevate them to top level categories. We can make any grouping we think where a top level discussion category might also have a place, in Discourse logic.

I'm actually quite fond of the Beginner category. It's better for beginners to have something simple to turn to, because they probably don't even know yet where they should be asking their questions. It's the same way the #beginners channel on Discord is nice and simple and helps people. Having this category and the corresponding channel on Discord was one of the main drivers to start these in the first place, after all.

For announcements, wouldn't it make more sense to have them as the first subcategory of the main category they are announcements for, just like we did with MLIR? So that people who look at it immediately see is as the first thing in the category.

**Comment by tstellar** [comment]
@Somae I had the same concern you did that a lot of top-level categories would make the main page too cluttered, but after looking at the current landing page, I don't think it will be a problem after all, but this probably depends a lot on personal preference.

You make a good point about posting in the main category.  For me, the fact that you can post in the main category when sub-categories exist is confusing.  Is there some way to disallow posting in the main category?

**Comment by tlattner** [comment]
I believe that Announcements should be a top level category and at the top of the list. Not buried within something else. This would include things such as releases, blog posts (so people can have discussion), dev mtg announcements. It avoids the need to cross post to different "lists" like we do now and easy for people to find. A good example of this is on the Swift Forums. 

I agree that having every sub-project as its own top level category might be a lot. I was asking more what the criteria would be that a project has its own top level category. We should define that to make it more clear. LLVM is a large project and while that might make the forums bigger by having many top level categories, we should do that if projects require the subcategories and it makes sense.  I think lumping everything into LLVM Project is a lot. It also makes it seem like MLIR and anything else outside of it is not a part of the LLVM Project.  

Right now, the cfe-users list has low subscriptions and low posts. In my opinion, the nature of the mailing list setup caused this problem (separation due to having to sign up for multiple lists). Discourse will hopefully help solve it and help users get the answers they need and developers be able to see the posts more easily. I think we should think carefully about how to get the most engagement and help for users of Clang versus the current setup. This applies to other projects as well. 

I'm not anti-Beginners category but I'm just wondering how successful it will be when we move everything over. LLVM Project is really large. Will the right people be reading and responding to beginners? Will people be able to easily filter out beginner questions to their area of expertise? Would it make more sense to have a Beginners category in sub-projects?

**Comment by joker-eph** [comment]
> I'm not anti-Beginners category but I'm just wondering how successful it will be when we move everything over. LLVM Project is really large. Will the right people be reading and responding to beginners? Will people be able to easily filter out beginner questions to their area of expertise? Would it make more sense to have a Beginners category in sub-projects?

We kind of already have this problem on Discord, for example beginner in MLIR may be asking questions in the beginner channel, but because no one will answer they will be redirected to the MLIR channel (where we welcome beginners of course! ;) ).

The same may exist for lld or other subproject: the people willing to help beginners with lld may not want to triage the beginner channel when most of the requests are for llvm or clang?

**Comment by tlattner** [comment]
> > I'm not anti-Beginners category but I'm just wondering how successful it will be when we move everything over. LLVM Project is really large. Will the right people be reading and responding to beginners? Will people be able to easily filter out beginner questions to their area of expertise? Would it make more sense to have a Beginners category in sub-projects?
> 
> We kind of already have this problem on Discord, for example beginner in MLIR may be asking questions in the beginner channel, but because no one will answer they will be redirected to the MLIR channel (where we welcome beginners of course! ;) ).
> 
> The same may exist for lld or other subproject: the people willing to help beginners with lld may not want to triage the beginner channel when most of the requests are for llvm or clang?

This is good information to know and something we should really consider what is the best Beginner experience and what will result in their questions being answered.

**Comment by tstellar** [comment]
I think we should resist creating separate developer and user sub-categories unless it's for a very high-traffic category.  For me, I think clang is probably the only category that needs the split.  You could probably make the case to have a separate user sub-category for llvm, but what is an llvm user?  A library user? A user of the bintuils replicas?  I don't really know.

**Comment by Somae** [comment]
I agree that there should be more top level categories, and also +1 for renaming the "LLVM Project" top category as it can be quite misleading. As long as we find a good balance of number of top level categories, I think it will be a great improvement to what we have right now. When elevating some subprojects to a top level category, maybe it is worth thinking about if more subcategories would make sense? Some mailing lists are quite high traffic now, and as soon as people don't have to subscribe to different mailing lists anymore to follow various topics, it might be worth splitting some "dev" up, for example. What are your thoughts? On Discord for example we have a more fine grain channel distribution right now.

It seems like we could take away posting rights from users for top level categories. We can't really make them just structural elements - they will continue to be forums.

The Beginner topic is a really hard one. It would be great to not have a place where some people have to triage, but that might result in people turning away who don't know where to ask their question, such increasing the barrier to asking the question in the first place. It's also not feasible for everyone to triage Beginners all the time. As long as redirecting or pinging members from the various communities works, it doesn't seem like the worst option either. Helping beginner triage could be a way to support the LLVM community, and we could make it so that topics are moved into the right category, for example, or something similar. Just trying to brainstorm here.

**Comment by tstellar** [comment]
I want to focus on the 'project' categories, since we need to get the mailman -> Discourse mappings done in order to start the migration.  I think for new topics like Beginner, we have more time to discuss and try to get right.

Here is a revised proposal for the project categories with less top-level and overall categories than my previous proposal.  What do you think?

- clang (Developer discussion on the top-level category)
  - users
  - clangd
- llvm
  - lld 
- runtimes
  - libcxx (includes parallel-libs, libunwind).
  - sanitizers
  - openmp
  - libclc
- flang
- lldb

Once we get this set, the last thing we need to complete the mappings is the announce /release-testers categories.

**Comment by Somae** [comment]
I think that looks like a really reasonable proposal, Tom! Thank you for putting so much consideration into this. Flang and LLDB can reasonably stand on their own, and maybe they'd like to have some sort of sub-categories at some point.

I'm happy to implement those changes on the Discourse end. Should we use a logo for some of these top level categories? If so, which specific ones?

**Comment by tlattner** [comment]
@lattner Adding Chris to this thread.

Sorry to drag this out. I think we are heading in the right direction. I can understand wanting to just focus on projects and the lists, but I see this as an opportunity to make some changes. I see 2 main problems from our current mailing list setups. 1) Lists like LLVM-dev are a firehose and need to be split  2) Making sure the right people are there to answer the questions (ie. users, beginners, etc). I think we should design the discourse categories to solve as much of the problems as we can and not to design them around the mistakes we have made in the past. This may mean that a list doesn't get mapped to a new discourse category. For example, I would like to propose that we put LLVM dev into an LLVM Dev List Archive category and then any recent and important posts from the import can be moved to the right category. 

I would also suggest that we not limit our category names to one word. :) 

I would want to see something like

Announce (to solve the cross posting issue and have a controlled and low volume category)
Clang Compiler & Tools
  - Using Clang
  - Building Clang
  - clangd
LLVM Based Subprojects
- lld
- flang (I would put flang here unless its big enough to warrant a whole category)
- lldb (may need top level but haven't looked at the traffic there)
- Static Analyzer (should we have one for this?)
LLVM Code Generation
- (we should have subcategories for all the major backends)
- Other (named something better but for smaller backends)
LLVM IR & Optimizations
- Sub categories such as a loop optimizations, etc
LLVM Dev List Archives
Community
Infrastructure
- Policies (we should add this as a subcategory to discuss project wide policies)
Runtimes
- Your list is good


I'm also coming around to keeping Beginners as I don't love having one in each major category. I think we should consider moving questions more aggressively to the right category when it makes sense to encourage the right people to respond.

**Comment by tlattner** [comment]
Let's try this again as the formatting got removed. Also, after thinking more,I think relying on tags is probably not a good approach as no one is forced to use them versus picking a category, they are. 

* Announce (to solve the cross posting issue and have a controlled and low volume category)
* Clang Compiler & Tools
    * Using Clang
    * Building Clang
    * clangd
* LLVM Based Subprojects
    * lld
    * flang (I would put flang here unless its big enough to warrant a whole category)
    * lldb (may need top level but haven't looked at the traffic there)
    * Static Analyzer (should we have one for this?)
* LLVM Code Generation
    * X86, ARM, (we should have subcategories for all the major backends)
    * Other (named something better but for smaller backends)
* LLVM IR & Optimizations
    * Sub categories such as a loop optimizations, etc
* LLVM Dev List Archives
* Community
    * Keep what we have, add as needed 
* Infrastructure
    * Add a Policies category (we should add this as a subcategory to discuss project wide policies)
* Runtimes
    * Your list is good

**Comment by joker-eph** [comment]
> I'm also coming around to keeping Beginners as I don't love having one in each major category. I think we should consider moving questions more aggressively to the right category when it makes sense to encourage the right people to respond.

One drawback of posting to a beginner category and then moving threads to the relevant project may be that the beginner category could look "empty" and "seemingly inactive" to a newcomer, who may then hesitate to post a question there (sorry I don't have a clear idea on all the tradeoff to make another suggestion though).

Otherwise I like your current proposal :)

**Comment by tstellar** [comment]
@tlattner I like this proposal too.  For the "LLVM Dev List Archives" category, does this mean we will archive the existing mailing lists into sub-categories here instead of trying to archive them into the new categories?

**Comment by tlattner** [comment]
I just mean for llvm-dev, there isn’t a good place to put LLVM-dev archives if we create more top level categories for LLVM. I am suggesting that we do make these top level categories as LLVM core is so big. 

As for the other lists, most can be dumped into the corresponding category/subcategory. Does that make sense? Pretty much what you already have suggested. I think I was saying keep Flang and LLDB as subcategories but we can debate that too….

We could do some archive for all of them but I think that messier…


-Tanya

**Comment by tstellar** [comment]
I've updated the mappings list based on Tanya's proposal.  The only one we haven't really discussed yet is the release-testers list.  I put in the patch that this would map to community/releases, but I'm open to other suggestions.

**Comment by tstellar** [comment]
@tlattner Can you take a quick look at the final list of mailing lists -> Discourse categories and confirm it looks OK, so I can send it off to the Discourse team?

File: discourse_migration/userguide.md [hunk 1]
```diff
@@ -45,48 +45,49 @@ are the typical use cases:
 
 This table explains the mapping from mailing lists to categories in Discourse.
 The email addresses of these categories will remain the same, after the
+migration.  Deprecated lists will become read-only as part of the Discourse
```
Comment (lattner) [inline discourse_migration/userguide.md:48]: "Deprecated" implies that they will be replaced with some new thing, and are limping along until that happens.

How about "Obsolete" or "Frozen" as names?

File: discourse_migration/userguide.md [hunk 2]
```diff
@@ -45,48 +45,49 @@ are the typical use cases:
...
+| cfe-dev               | Clang Compiler & Tools |
+| cfe-users             | Clang Compiler & Tools/Using Clang |
+| clangd-dev            | Clang Compiler & Tools/clangd |
+| devmtg-organizers     | Deprecated |
 | Docs                  | infrastructure/documentation |
-| eurollvm-organizers   | **TODO** create new category |
+| eurollvm-organizers   | Deprecated |
 | flang-commits         | no migration at the moment |
-| flang-dev             | llvm-project/flang |
+| flang-dev             | LLVM Based Subprojects/flang |
```
Comment (lattner) [inline discourse_migration/userguide.md:63]: I'd recommend naming this just "Subprojects / Flang".  No need for "LLVM Based".  Similarly, I think we should rename "LLVM Incubator" to just "Incubator".  The extra verbosity shows up in email subject lines, like this:

`Subject: [LLVM] [LLVM Incubator/CIRCT] Proposal for Per-instance Metadata`

It would be better to have:

`Subject: [LLVM] [Incubator/CIRCT] Proposal for Per-instance Metadata`
`Subject: [LLVM] [Subproject/Flang] Fancy fortran topic`

I'd also recommend capitalizing "Flang" in the topic name just like we do "Clang" for consistency.  It is talking about the subproject, not the executable.

File: discourse_migration/userguide.md [hunk 3]
```diff
@@ -45,48 +45,49 @@ are the typical use cases:
...
 | gsoc                  | community/gsoc |
 | libc-commits          | no migration at the moment |
-| libc-dev              | llvm-project/libc |
-| Libclc-dev            | **TODO** create new category |
+| libc-dev              | Runtimes/libc |
+| Libclc-dev            | Runtimes/libclc |
```
Comment (lattner) [inline discourse_migration/userguide.md:67]: What do you think of the alternative:

`Runtimes/C - libc`
`Runtimes/C++ - libcxx`
`Runtimes/OpenCL - libclc`

etc?  That will be more discoverable/understandable for new people.

File: discourse_migration/userguide.md [hunk 4]
```diff
@@ -45,48 +45,49 @@ are the typical use cases:
...
 | libcxx-bugs           | no migration at the moment |
 | libcxx-commits        | no migration at the moment |
-| libcxx-dev            | llvm-project/libcxx |
+| libcxx-dev            | Runtimes/libcxx |
 | lldb-commits          | no migration at the moment |
-| lldb-dev              | llvm-project/lldb |
-| llvm-admin            | **TODO** create new category |
-| llvm-announce         | **TODO** create new category |
+| lldb-dev              | LLVM Based Subprojects/lldb |
+| llvm-admin            | Replace with bugzilla or Github Issues |
```
Comment (lattner) [inline discourse_migration/userguide.md:73]: Recommend "no migration at the moment" unless there is a specific plan already.

File: discourse_migration/userguide.md [hunk 5]
```diff
@@ -45,48 +45,49 @@ are the typical use cases:
...
+| llvm-announce         | Announce |
```
Comment (lattner) [inline discourse_migration/userguide.md:74]: I recommend "Community / Announcements"

File: discourse_migration/userguide.md [hunk 6]
```diff
@@ -45,48 +45,49 @@ are the typical use cases:
...
 | Openmp-commits        | no migration at the moment |
-| Openmp-dev            | llvm-project/openmp |
+| Openmp-dev            | Runtimes/openmp |
 | Parallel_libs-commits | no migration at the moment |
-| Parallel_libs-dev     | **TODO** use llvm-project/pstl or create new category ? |
-| Release-testers       | **TODO** create new category |
-| Test-list             | **TODO** create new category |
-| vmkit-commits         | no migration at the moment |
+| Parallel_libs-dev     | Runtimes/libcxx |
+| Release-testers       | community/releases |
```
Comment (lattner) [inline discourse_migration/userguide.md:86]: "Community / Releases"  capitalized plz

File: discourse_migration/userguide.md [hunk 7]
```diff
@@ -45,48 +45,49 @@ are the typical use cases:
...
+| Test-list             | Deprecated |
+| vmkit-commits         | Deprecated |
 | WiCT                  | community/women-in-compilers-and-tools/ |
```
Comment (lattner) [inline discourse_migration/userguide.md:89]: This is using the URL format instead of the name "Community / Women in Compilers and Tools".  It would be good to be consistent one way or the other.

**Review by lattner — CHANGES_REQUESTED** [review]

**Comment by lattner** [comment]
A couple other questions:
 - Will this also remove/replace the  existing "LLVM Project" section, or is that a second step?
 - Is it possible to move this from `https://llvm.discourse.group/` to the more conventional `forums.llvm.org` URL?

**Comment by lattner** [comment]
Ah, I just saw Tanya's comment above.  I think it would be great to have a "Target Specific" category, then "X86, ARM, RISC-V, ..., Other" underneath it.  We can't really auto-populate that right now, but those will likely have their own different interest groups and it makes total sense to carve that out as a top level thing in the forum, given it is orthogonal to the frontends and runtimes.

File: discourse_migration/userguide.md (hunk 5 above)
Comment (tlattner) [reply discourse_migration/userguide.md:74]: I would really prefer a top level Announcement category. The goal is that it is featured (top of page) and that people can only subscribe to this one thing and it is very easy to find, etc. I really like how Swift has this at the top and then the cross post the blog posts there so people can comment. Community also seems to be a bucket that could be considered "not technical" and we want to release announcements over the whole project as well.

File: discourse_migration/userguide.md (hunk 4 above)
Comment (tlattner) [reply discourse_migration/userguide.md:73]: No migration for LLDB?

File: discourse_migration/userguide.md (hunk 3 above)
Comment (tlattner) [reply discourse_migration/userguide.md:67]: I agree on this. Much easier to understand.

File: discourse_migration/userguide.md (hunk 2 above)
Comment (tlattner) [reply discourse_migration/userguide.md:63]: I do think we should consider if most people are really using email though when deciding how to name things. This particular case seems fine though.

File: discourse_migration/userguide.md (hunk 5 above)
Comment (lattner) [reply discourse_migration/userguide.md:74]: Fair point.  A top level "Announcements" category (not "Announce" plz) SGTM!

File: discourse_migration/userguide.md (hunk 7 above)
Comment (tlattner) [reply discourse_migration/userguide.md:89]: Agreed. I think without the dashes.

File: discourse_migration/userguide.md (hunk 3 above)
Comment (tstellar) [reply discourse_migration/userguide.md:67]: > What do you think of the alternative:
> 
> `Runtimes/C - libc`
> `Runtimes/C++ - libcxx`
> `Runtimes/OpenCL - libclc`
> 
> etc? That will be more discoverable/understandable for new people.
 
So your suggestion is to use "Runtimes/C - libc" as the full category name with no sub-categories?

Also, what about OpenMP and sanitizers.  We had those as sub-categories in the Runtime category.  Should those also become top-level categories?

**Comment by tlattner** [comment]
I think the main thing left is Release Testing. I was sorta leaning towards having it being a top level category and then sub topics could be per-release type? Maybe this is overkill? Community just doesn't feel like the right spot for it though and its cross project. I do think we can put it under Community and move it out later, but not positive.

File: discourse_migration/userguide.md (hunk 3 above)
Comment (tlattner) [reply discourse_migration/userguide.md:67]: I read it as 
Runtimes/C
Runtimes/C++
Runtimes/OpenCL 

And then the - name was just showing what it mapped to. Maybe I misunderstood.

Comment (lattner) [reply discourse_migration/userguide.md:67]: I meant a top level "Runtimes" category and then "C - libc" as a subcategory.  I think that putting "OpenMP" and "Sanitizers" as subcategory of Runtimes would also make a lot of sense.

My understanding is that IWG is focused on getting the mailing lists mapped out.  Once the base structure is in place (e.g. the top level Runtimes category) we can fill in the new entries on demand.

File: discourse_migration/userguide.md (hunk 4 above)
Comment (lattner) [reply discourse_migration/userguide.md:73]: This is proposing LLDB go into "Subprojects / LLDB"

File: discourse_migration/userguide.md (hunk 3 above)
Comment (lattner) [reply discourse_migration/userguide.md:67]: Incidentally, I think it would also be worth going simpler than this.  The name of the runtime library is confusing in many cases.  I think it is worth considering just doing:

```
Runtimes / C
Runtimes / C++
Runtimes / OpenCL
Runtimes / OpenMP
Runtimes / Sanitizers
Runtimes / Other
```

Comment (tstellar) [reply discourse_migration/userguide.md:67]: Could we have the top-level 'Runtimes' category be for posts about other runtimes, or do you think it's better to have a separate sub-category?

Comment (lattner) [reply discourse_migration/userguide.md:67]: Both work.  The downside of reusing the top level for this is that you can't follow "just that".  You'd have to follow the entire "Runtimes" category and you'd get email about everything.  This would be annoying if you just care about polly or something, you'd rather just follow "Runtimes/Other".

Comment (lattner) [reply discourse_migration/userguide.md:67]: btw, it isn't clear to me that even "Sanitizers" will be high volume.  I think we should either start with a minimal set of sub categories (just to make the mailing list transition work) and add on demand, or go full in and add subcategories for all the things we know of.

**Comment by tstellar** [comment]
I think I've applied all the suggestions now.  I also added a file that lists all the categories and sub-categories since it was getting a little hard to follow in the comments.  Please take a look and let me know what you think.

File: discourse_migration/discourse-categories.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+* Releases
+
+* Clang Compiler & Tools
+  - Using Clang
+  - Building Clang
+  - clangd
+
+* Subprojects
+  - lld
```
Comment (lattner) [inline discourse_migration/discourse-categories.md:13]: I'd recommend "LLD Linker"

File: discourse_migration/discourse-categories.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
+# Proposed Discourse Categories
+
+* Announcements
+
+* Releases
```
Comment (lattner) [inline discourse_migration/discourse-categories.md:5]: Will this have anything other than announcements, how about putting Releases as a subcategory under Announcements?

If it isn't announce'y it would be good to put it somewhere else - we have a bunch of top-levels

File: discourse_migration/discourse-categories.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+* Clang Compiler & Tools
+  - Using Clang
+  - Building Clang
+  - clangd
+
+* Subprojects
+  - lld
+  - Flang
+  - lldb
```
Comment (lattner) [inline discourse_migration/discourse-categories.md:15]: I'd recommend "LLDB Debugger"

File: discourse_migration/discourse-categories.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+  - Static Analyzer
+
+* Code Generation
+  - AArch64
```
Comment (lattner) [inline discourse_migration/discourse-categories.md:19]: I'd suggest "Common Infra" or something similar as a subcategory.

File: discourse_migration/discourse-categories.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+  - AMDGPU
+  - ARC
+  - ARM
+  - AVR
+  - BPF
+  - CSKY
+  - Hexagon
+  - Lanai
+  - M68k
```
Comment (lattner) [inline discourse_migration/discourse-categories.md:28]: We have a lot of backends, and some like Lanai, M68k, SystemZ are niche and will have very little public traffic.  I'd recommend we start with 8 or 10 of the big and common ones, and then have an other category.  We can "promote" things out of other when there is a lot of traffic in the Other category.

File: discourse_migration/discourse-categories.md [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+  - VE
+  - WebAssembly
+  - X86
+  - XCore
+
+* IR & Optimizations
+  - Loop Optimizations
+
+* Community
+  - Women in Compilers and Tools
```
Comment (lattner) [inline discourse_migration/discourse-categories.md:45]: Should WICT and Community.o get merged?

File: discourse_migration/discourse-categories.md [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+  - Job Postings
+  - US Developer Meeting
+  - EuroLLVM
```
Comment (lattner) [inline discourse_migration/discourse-categories.md:48]: Should we merge EuroLLVM and US DevMtg?

File: discourse_migration/discourse-categories.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+  - GSoC
```
Comment (lattner) [inline discourse_migration/discourse-categories.md:49]: I'd recommend "Google Summer of Code"

File: discourse_migration/discourse-categories.md [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+  - Community.o
+  - Foundation
```
Comment (lattner) [inline discourse_migration/discourse-categories.md:51]: I know it is inconsistent, but I'd recommend "LLVM Foundation" since that's a proper name.

File: discourse_migration/discourse-categories.md [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+
+* Project Infrastructure
+  - Website
+  - Documentation
+  - GitHub
+  - Code Review
+  - Discord
+  - Mailing Lists
```
Comment (lattner) [inline discourse_migration/discourse-categories.md:59]: Should "Mailing Lists" be "Mailing Lists and Forums"?

File: discourse_migration/discourse-categories.md [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+  - C++
+  - OpenCL
+  - OpenMP
+  - Sanitizers
+
+* Incubator
+  - CIRCT
+  - mlir-npcomp
+
+* LLVM Dev List Archives
```
Comment (lattner) [inline discourse_migration/discourse-categories.md:75]: top level items get a lot of real estate on the front page, and this is only for historical reference / searching. Can this be "Community / Archived llvm-dev List" or something like that?

File: discourse_migration/discourse-categories.md [hunk 19]
```diff
@@ -0,0 +1,75 @@
+# Proposed Discourse Categories
+
+* Announcements
+
+* Releases
+
+* Clang Compiler & Tools
```
Comment (lattner) [inline discourse_migration/discourse-categories.md:7]: This is pretty verbose, how about `Clang Frontend`?

File: discourse_migration/discourse-categories.md [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+  - Using Clang
+  - Building Clang
+  - clangd
+
+* Subprojects
+  - lld
+  - Flang
```
Comment (lattner) [inline discourse_migration/discourse-categories.md:14]: "Flang Fortran Frontend"?

File: discourse_migration/discourse-categories.md (hunk 12 above)
Comment (lattner) [reply discourse_migration/discourse-categories.md:28]: As a starting point, I'd go with:

"AArch64", "ARM", "Mips", "PowerPC", "RISCV", "WebAssembly", "X86", and "Other".

Rationale that those are the ones (I'm aware of) that have more than one organization as the primary driver and that are moderately popular.  This will be a new thing for the community, so we can always add more if people squawk loudly.

File: discourse_migration/discourse-categories.md (hunk 9 above)
Comment (tstellar) [reply discourse_migration/discourse-categories.md:5]: The Release  category would have the discussion among release testers which currently takes place on the release-testers list.  I also get a lot of questions about the releases asked to me either by direct email or on IRC/Discord, so my hope is that this category could be used for those.

File: discourse_migration/discourse-categories.md (hunk 11 above)
Comment (tstellar) [reply discourse_migration/discourse-categories.md:19]: Can we use the top-level category for that?

File: discourse_migration/discourse-categories.md (hunk 9 above)
Comment (lattner) [reply discourse_migration/discourse-categories.md:5]: Ok, I'd recommend moving this under "Project Infrastructure" category then.

File: discourse_migration/discourse-categories.md [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+* Clang Frontend
+  * Using Clang
+  * Building Clang
+  * clangd
+
+* Subprojects
+  * LLD Linker
+  * Flang Fortran Frontend
+  * LLDB Debugger
+  * Static Analyzer
```
Comment (lattner) [inline discourse_migration/discourse-categories.md:16]: I'd move "Static Analyzer" under "Clang Frontend" since it is specific to that.

**Review by lattner — APPROVED** [review]
Looks great!  (a couple more suggestions inline)

**Comment by lattner** [comment]
LGTM Tom!

File: discourse_migration/discourse-categories.md (hunk 9 above)
Comment (tlattner) [reply discourse_migration/discourse-categories.md]: I'm not really sure it should go under project infrastructure. It really has nothing to do with infrastructure. What is the negative with it being a top level category?

File: discourse_migration/discourse-categories.md (hunk 19 above)
Comment (tlattner) [reply discourse_migration/discourse-categories.md]: If we are naming this Clang Frontend, why are we putting other tools here then too? Should they just go under Subprojects then?

File: discourse_migration/discourse-categories.md (hunk 18 above)
Comment (tlattner) [reply discourse_migration/discourse-categories.md]: I'm not loving this being under Community. I would rather it be under Project Infra if we aren't going to make it top-level. Or stick it in IR or something. Top level seems more appropriate though given it fits nothing of the top level categories exactly.

Comment (lattner) [reply discourse_migration/discourse-categories.md]: "Project Infra" or "IR" would also be fine to me.   Top Level doesn't make sense because it will get a LOT of space on the [front page](https://llvm.discourse.group/) and be literally useless (in terms of not something anyone would click on).

File: discourse_migration/discourse-categories.md (hunk 11 above)
Comment (lattner) [reply discourse_migration/discourse-categories.md]: My thought is that "Code Generation / Common Infra" would be things like SelectionDAG, Global ISel, TargetRegisterInfo, etc.  All Code generation related things.

File: discourse_migration/discourse-categories.md (hunk 9 above)
Comment (lattner) [reply discourse_migration/discourse-categories.md]: Top level items take a lot of visible area on the [main page](https://llvm.discourse.group/).  Discussion between release testers seems extremely infra, or pontentially "community" level, not top level.

Comment (lattner) [reply discourse_migration/discourse-categories.md]: Actually, here's what I'd recommend doing: I'd suggest renaming this to "Release Testing" and put it under Infrastructure.  If we want release announcements to be separate from general announcements then we can have a separate "Announcements / Releases" subcategory for them.

**Comment by tstellar** [comment]
Summary of outstanding questions:
- What to do about Release category?  (Latest suggestion is Infrastructure/Release Testing)
- Should we merge WICT and Community.o ?
- Should we merge EuroLLVM and US DevMtg?
- Should we move tools out of "Clang Frontend" and into "Subprojects"

**Review by lattner — APPROVED** [review]

File: discourse_migration/userguide.md [hunk 22]
```diff
@@ -45,48 +45,49 @@ are the typical use cases:
...
-| cfe-users             | **TODO** llvm-project/clang or create new category? |
-| clangd-dev            | llvm-project/clangd |
-| devmtg-organizers     | **TODO** create new category |
-| Docs                  | infrastructure/documentation |
-| eurollvm-organizers   | **TODO** create new category |
+| cfe-dev               | Clang Frontend |
+| cfe-users             | Clang Frontend/Using Clang |
+| clangd-dev            | Clang Frontend/clangd |
+| devmtg-organizers     | Obsolete |
+| Docs                  | Project Infrastructure/Documentation |
```
Comment (tlattner) [inline discourse_migration/userguide.md:60]: Docs should be Obsolete

File: discourse_migration/userguide.md [hunk 23]
```diff
@@ -45,48 +45,49 @@ are the typical use cases:
...
+| eurollvm-organizers   | Obsolete |
 | flang-commits         | no migration at the moment |
-| flang-dev             | llvm-project/flang |
-| gsoc                  | community/gsoc |
+| flang-dev             | Subprojects/Flang |
```
Comment (tlattner) [inline discourse_migration/userguide.md:63]: Just a note it should be - Flang Fortran Frontend

File: discourse_migration/userguide.md [hunk 24]
```diff
@@ -45,48 +45,49 @@ are the typical use cases:
...
+| gsoc                  | Community/Google Summer of Code |
```
Comment (tlattner) [inline discourse_migration/userguide.md:64]: Obsolete  (no real archives)

File: discourse_migration/userguide.md (hunk 5 above)
Comment (tlattner) [reply discourse_migration/userguide.md:74]: Please update to Announcements

File: discourse_migration/userguide.md [hunk 25]
```diff
@@ -45,48 +45,49 @@ are the typical use cases:
...
+| llvm-announce         | Announce |
 | llvm-branch-commits   | no migration at the moment |
 | llvm-bugs             | no migration at the moment |
 | llvm-commits          | no migration at the moment |
-| llvm-dev              | llvm-project/llvm |
-| llvm-devmeeting       | community/devmtg |
-| llvm-foundation       | **TODO**  create new category|
+| llvm-dev              | Project Infrastructure/LLVM Dev List Archives |
+| llvm-devmeeting       | Community/US Developer Meeting |
+| llvm-foundation       | Community/Foundation |
```
Comment (tlattner) [inline discourse_migration/userguide.md:80]: Change to LLVM Foundation
